// next-intl/dist/development/shared/utils.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function isRelativeHref(href) {
  const pathname = typeof href === 'object' ? href.pathname : href;
  return pathname != null && !pathname.startsWith('/');
}
function isLocalHref(href) {
  if (typeof href === 'object') {
    return href.host == null && href.hostname == null;
  } else {
    const hasProtocol = /^[a-z]+:/i.test(href);
    return !hasProtocol;
  }
}
function isLocalizableHref(href) {
  return isLocalHref(href) && !isRelativeHref(href);
}
function localizeHref(href, locale) {
  let curLocale = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : locale;
  let curPathname = arguments.length > 3 ? arguments[3] : undefined;
  let prefix = arguments.length > 4 ? arguments[4] : undefined;
  if (!isLocalizableHref(href)) {
    return href;
  }
  const isSwitchingLocale = locale !== curLocale;
  const isPathnamePrefixed = hasPathnamePrefixed(prefix, curPathname);
  const shouldPrefix = isSwitchingLocale || isPathnamePrefixed;
  if (shouldPrefix && prefix != null) {
    return prefixHref(href, prefix);
  }
  return href;
}
function prefixHref(href, prefix) {
  let prefixedHref;
  if (typeof href === 'string') {
    prefixedHref = prefixPathname(prefix, href);
  } else {
    prefixedHref = {
      ...href
    };
    if (href.pathname) {
      prefixedHref.pathname = prefixPathname(prefix, href.pathname);
    }
  }
  return prefixedHref;
}
function unprefixPathname(pathname, prefix) {
  return pathname.replace(new RegExp("^".concat(prefix)), '') || '/';
}
function prefixPathname(prefix, pathname) {
  let localizedHref = prefix;

  // Avoid trailing slashes
  if (/^\/(\?.*)?$/.test(pathname)) {
    pathname = pathname.slice(1);
  }
  localizedHref += pathname;
  return localizedHref;
}
function hasPathnamePrefixed(prefix, pathname) {
  return pathname === prefix || pathname.startsWith("".concat(prefix, "/"));
}
function matchesPathname( /** E.g. `/users/[userId]-[userName]` */
template, /** E.g. `/users/23-jane` */
pathname) {
  const regex = templateToRegex(template);
  return regex.test(pathname);
}
function getLocalePrefix(locale, localePrefix) {
  var _localePrefix$prefixe;
  return localePrefix.mode !== 'never' && ((_localePrefix$prefixe = localePrefix.prefixes) === null || _localePrefix$prefixe === void 0 ? void 0 : _localePrefix$prefixe[locale]) ||
  // We return a prefix even if `mode: 'never'`. It's up to the consumer
  // to decide to use it or not.
  '/' + locale;
}
function templateToRegex(template) {
  const regexPattern = template
  // Replace optional catchall ('[[...slug]]')
  .replace(/\[\[(\.\.\.[^\]]+)\]\]/g, '?(.*)')
  // Replace catchall ('[...slug]')
  .replace(/\[(\.\.\.[^\]]+)\]/g, '(.+)')
  // Replace regular parameter ('[slug]')
  .replace(/\[([^\]]+)\]/g, '([^/]+)');
  return new RegExp("^".concat(regexPattern, "$"));
}

exports.getLocalePrefix = getLocalePrefix;
exports.hasPathnamePrefixed = hasPathnamePrefixed;
exports.isLocalHref = isLocalHref;
exports.isLocalizableHref = isLocalizableHref;
exports.isRelativeHref = isRelativeHref;
exports.localizeHref = localizeHref;
exports.matchesPathname = matchesPathname;
exports.prefixHref = prefixHref;
exports.prefixPathname = prefixPathname;
exports.templateToRegex = templateToRegex;
exports.unprefixPathname = unprefixPathname;
